src/metabase/models/serialization/base.clj

```diff
@@ -0,0 +1,171 @@
+    this performs whatever steps are necessary to update or insert the incoming entity.
+    Called for each **file**, which usually means for each entity. For single-file output like for settings, this is
+    called once for the whole file.
+    Defaults to [[deserialize-file-plus]] with [[identity]] as the inner transformation.")
+
+  (deserialize-upsert
+    [old-entity new-map]
+    "Given the original entity and the new deserialized map (not an IModel yet), upsert appdb to merge in the updated
+    entity.
+    Defaults to a naive update by primary key, using just the new map's values.")
+
+  (deserialize-insert
+    [this new-map]
+    "Given the model (eg. Collection) and the new deserialized map (not an IModel), insert this new record into
+    the appdb.
+    Defaults to a straight db/insert! of this new map."))
+
+(defn- default-serialize-all [model user-or-nil]
+  (eduction (map serialize-one) (serialize-query model user-or-nil)))
+
+(defn- default-query [model _]
+  (db/select-reducible model))
+
+(defn serialize-one-plus
+  "Helper for applying the usual serialization steps to a single entity, followed by a user-supplied function with any
+  specific logic the entity needs."
+  [transform named]
+  (fn [entity]
+    [(format "%s%s%s%s.yaml"
+             (name entity)
+             File/separatorChar
+             (or (:entity_id entity)
+                 (serdes.hash/identity-hash entity))
+             (when named
+               (str "+" (named entity))))
+     (-> (into {} entity)
+         (dissoc (models/primary-key entity))
+         (assoc :serdes_type (name entity))
+         transform)]))
+
+(defn- default-upsert [old-entity new-map]
+  (db/update! (symbol (name old-entity)) (get old-entity (models/primary-key old-entity)) new-map)
+  #_(prn "upsert" old-entity new-map))
+
+(defn- default-insert [model new-map]
+  (db/simple-insert! model new-map)
+  #_(prn "insert" new-map))
+
+(defn deserialize-file-plus
+  "Given a function, this returns a function suitable for use as [[deserialize-file]]. The provided function is applied
+  to the EDN map read from the YAML file, and then either [[deserialize-upsert]] or [[deserialize-insert]] is called
+  depending on whether a corresponding entity was found in the database."
+  [f]
+  (fn [model contents local]
+    (let [hydrated (f contents)]
+      (if local
+        (deserialize-upsert (db/select-one model :id local) hydrated)
+        (deserialize-insert model hydrated)))))
+
+(def ISerializableDefaults
+  "Default implementations for [[ISerializable]], so models need only override those that need special handling."
+  {:serialize-all         default-serialize-all
+   :serialize-query       default-query
+   :serialize-one         (serialize-one-plus identity nil)
+   :deserialize-file      (deserialize-file-plus identity)
+   :deserialize-upsert    default-upsert
+   :deserialize-insert    default-insert})
```

Comment:
I would vote making these multimethods instead. Then you don't need to muck around with `ISerializableDefaults` hacks and stuff like that. It's more idiomatic too.

(I know Toucan does things this way but it's a dumb way of doing things and I wouldn't do it that way again. Toucan 2 uses multimethods for stuff like `pre-insert` and `post-select`.)

Reply:
I've made it so with a major refactoring. I like the way the multimethods work!